Add teaching cookbooks and materials

jdannberg opened the pull request:
I'd like to contribute my teaching materials to the ASPECT repository, so that others can use them, and so that the input files keep working with the development version. Here, I've only added a very first input file, so that we can discuss if the structure I am using works, or if we want to change something.

naliboff commented:
@jdannberg - Thanks for contributing this and I agree it would be great to have a collection of parameter files specifically for teaching.

A few questions regarding the organization:
1.  will `cookbooks/intro-geophysics` be only for the parameter files from the UF class or could other parameter files from similar classes elsewhere be added?
2. If the idea to make `cookbooks/intro-geophysics` specifically for the UF class, then maybe `cookbooks/intro-geophysics-uf` would be better?
3. For further clarity, what if we add an additional folder: `cookbooks/teaching/intro-geophysics`?

No strong opinions on any of the above questions, just thinking out loud :)

bangerth commented:
Just as one point of experience: The deal.II video lectures were originally numbered sequentially, but in the end more and more were added. They could easily fill a two-semester course now. In practice, what this means is that from a 3-day course to a semester-long course, I can now simply pick-and-choose a subset of these lectures (and so can others). I don't know whether @jdannberg is interested in others contributing to the teaching materials she uses for her classes, but I've had good experiences with a model where one (i) just adds more than may strictly be necessary and (ii) evolves and edits existing material, and concrete courses then boil down to selecting a subset of everything that's there.

jaustermann commented:
@jdannberg Juliane this is awesome! I like the 'pick and choose' approach that Wolfgang describes. Given that some of your material might build on each other, keeping a numbering scheme might be useful as a rough guideline for progression. It might also be useful to note in the parameter file what the suggested 'prerequisite' cookbooks are for this one. Are you thinking of also adding slides and / or recorded lectures? If so, maybe this might justify a whole separate 'teaching' folder outside of cookbooks. Looking forward to the meeting today to talk about this :)

jdannberg commented:
Alright, based on our discussion today, I made the cookbooks themselves normal cookbooks (without numbers), and I added a section in the manual for the course that can be used to list all cookbooks that would make up the course. 
This also removes the numbering. 

Does that structure look more reasonable?

tjhei commented on cookbooks/tutorial/doc/tutorial.tex:
```diff
@@ -0,0 +1,29 @@
+\subsubsection{Running a geodynamic model}
```
I am not sure about the name "tutorial"

tjhei commented on cookbooks/tutorial/tutorial.prm:
```diff
@@ -0,0 +1,122 @@
+include $ASPECT_SOURCE_DIR/cookbooks/convection-box/convection-box.prm
```
do you need this at all?

jdannberg replied:
Yes. I made a diff and then included only the things I changed. But I can also put this in as a completely separate file.

tjhei replied:
One could argue that this one might be the first one to teach, so it might be worth setting all parameters? Not sure.

tjhei commented on doc/manual/manual.tex:
```diff
@@ -10042,6 +10042,31 @@ \subsubsection{Brittle thrust wedges benchmark}
+  \item ...
+\end{enumerate}
+
+\input{cookbooks/tutorial/doc/tutorial.tex}
```
should these be put into the main section of cookbooks instead? This might be a good first cookbook or not?

jdannberg replied:
It may be a good input file for a first cookbook, but our current first cookbook (convection box) is almost the same, with less functionality, and that section contains a lot of helpful advice for first-time users. I thought the whole point about these teaching cookbooks is that the contributor doesn't have to write a lot of text in the manual, and instead there may be slides that accompany the cookbook. So I think this is somewhat different from the normal cookbooks.

tjhei commented on doc/manual/manual.tex:
```diff
@@ -10042,6 +10042,31 @@ \subsubsection{Brittle thrust wedges benchmark}
 \end{figure}
 
 
+\subsection{Setups for teaching}
```
If you put this .tex into a folder like ./cookbooks/teaching/intro-to-geophy/doc/, one could also add ./cookbooks/teraching/intro-to-geophy/README.md that you can point students to (with a simple ordered list like below).

jdannberg replied:
I know I probably wouldn't use a README like that (I think I may even just copy the relevant cookbooks to a separate git repo just for a class, in order to be able to add files easily; what I add to the manual is more meant for the instructors rather than the students). If you insist (or if others think it would be useful), I can add a readme, but for now I've simply linked the different cookbook folders in the manual, in addition to the section names.

naliboff commented on cookbooks/heat_flow/doc/heat-flow.tex:
```diff
@@ -0,0 +1,171 @@
+In your report, start by describing the setup of the model and what kind of predictions it can make that we can then compare to observations. Specifically, include a qualitative discussion of the physical processes and the material properties that affect the temperature distribution and heat flow computed in the model, based on the equation solved in the model. 
+
+Then describe your model results. Include your plot of convective/conductive heat flux and discuss it:
+Where is convection dominant, where conduction? 
```
```suggestion
Respectively, where are convection and conduction dominant? 
```

naliboff commented on cookbooks/heat_flow/doc/heat-flow.tex:
```diff
@@ -0,0 +1,171 @@
+
+Use the input file \url{cookbooks/heat_flow/heat-flow-plume.prm}. 
+The model is very similar to the mid-ocean ridge model you analyzed above, but the oceanic plate does not move
+(so by default, there is no convective heat flux), and the extension in vertical direction is larger 
```
```suggestion
(so by default, there is no convective heat flux), and extension in the vertical direction is larger 
```

naliboff commented on cookbooks/heat_flow/heat-flow-terms.prm:
```diff
@@ -0,0 +1,38 @@
+include $ASPECT_SOURCE_DIR/cookbooks/heat_flow/heat-flow.prm
```
Maybe worth adding a 1-2 sentences here for those that did not first read the relevant manual section?

naliboff commented on cookbooks/tutorial/doc/tutorial.tex:
```diff
@@ -0,0 +1,29 @@
+\label{sec:cookbooks-running-a-model}
+\textit{This section was contributed by Juliane Dannberg.}
+
+The input file for this model can be found at \url{cookbooks/tutorial/tutorial.prm}.
```
Perhaps change the name to t`utorial_convection_box_2d.prm` if other parameter files will be added to the folder?

gassmoeller replied:
I will let @jdannberg also comment, but I think we used `tutorial.prm` in a bunch of tutorials as the first file, and it would be nice to keep the name (so that old slides stll work). Also each parameter file will get its own folder, so as long as no other parameter file wants to claim the folder `tutorial` there should not be a conflict.

jdannberg replied:
That is a good point. I renamed it now to convection-box-particles, based on @tjhei's comments, but of course that means it's different in the slides now. But I could of course also update the slides.

jdannberg commented on cookbooks/heat_flow/doc/heat-flow.tex:
```diff
@@ -0,0 +1,171 @@
+\subparagraph{Heat flux close to mid-ocean ridges}
+Analyze the heat flux close to mid-ocean ridges. 
+
+In order to do that, use the mid-ocean ridge model we have been looking at in class, which is located in the folder \texttt{models/01\_heat\_flow/heat-flow.prm}. Run the model and use ParaView to compute and visualize the conductive and convective heat flux in vertical direction. 
```
Good catch! I just copied that from my assignment...

jdannberg commented:
@naliboff Thanks for your review! I hadn't expected a complete review (hence the WIP label), but thought it would be good to get the folder structure right before adding more cookbooks. But I addressed your comments, and of course I can always add my other cookbooks in a separate pull request.

jdannberg commented:
This pull request now contains 4 different cookbooks. I can add more in the future, but I think this is already a good starting point and we can add more cookbooks in follow-up PRs. 
I have rebased this and I have addressed all comments; so if someone wants to have a look at this now, that would be great!

naliboff commented on cookbooks/heat_flow/heat-flow.prm:
```diff
@@ -0,0 +1,122 @@
+    set Function expression = if(x>0, v * cm_per_year, -v * cm_per_year); 0
+  end
+end
+
```
Add a comment here describing what fixing the "y", but not "x" component of the boundary tractions represents and achieves?

naliboff commented on cookbooks/heat_flow/heat-flow.prm:
```diff
@@ -0,0 +1,122 @@
+  end
+end
+
+
```
```suggestion
# The initial temperature is constant throughout the model, but subsequently evolves through time.
```

naliboff commented on cookbooks/heat_flow/heat-flow.prm:
```diff
@@ -0,0 +1,122 @@
+  end
+end
+
+
```
Is there a previous cookbook that describes the composition reaction material model? Maybe a few brief sentences about this material model and the associated parameter choices?

naliboff commented on cookbooks/heat_flow/heat-flow.prm:
```diff
@@ -0,0 +1,122 @@
+
+
+# The final part of this input file describes how many times the
+# mesh is refined and what to do with the solution once computed
```
```suggestion
# mesh is refined and what to do with the solution once computed.
# Here, the mesh initially undergoes two global refinements and four
# adaptive refinements on the first time step, but is not refined further
# (i.e., static) in later model runs. The adaptive refinements use the 
# minimum refinement function to increase the resolution towards the
# model surface in a piece-wise fashion at depths of 0, 7000, 30000, and
# 45000 meters. 
```

naliboff commented on cookbooks/heat_flow/heat-flow.prm:
```diff
@@ -0,0 +1,122 @@
+end
+
+
+# The final part of this input file describes how many times the
```
```suggestion
# This part of this input file describes how many times the
```

naliboff commented on cookbooks/magnetic_stripes/CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+# Copyright (C) 2011 - 2019 by the authors of the ASPECT code.
```
```suggestion
# Copyright (C) 2011 - 2021 by the authors of the ASPECT code.
```

naliboff commented on cookbooks/magnetic_stripes/CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+
+CMAKE_MINIMUM_REQUIRED(VERSION 2.8.12)
+
+FIND_PACKAGE(Aspect 2.1.0 QUIET HINTS ${Aspect_DIR} ../ ../../ $ENV{ASPECT_DIR})
```
```suggestion
FIND_PACKAGE(Aspect 2.3.0 QUIET HINTS ${Aspect_DIR} ../ ../../ $ENV{ASPECT_DIR})
```

naliboff commented on cookbooks/magnetic_stripes/doc/magnetic_stripes.tex:
```diff
@@ -0,0 +1,38 @@
+\includegraphics[width=0.6\textwidth]{cookbooks/magnetic_stripes/doc/mid-ocean-ridge.png}
+\hfill
+\phantom.
+\caption{\it Setup of the mid-ocean ridge model. Background colors show temperature, black and white colors at the top of the model illustrate the orientation of the magnetic field frozen in the rock when the melt generated at the mid-ocean ridge reaches the surface, crystallizes to form new sea floor, and the rock cools down. White lines illustrate the flow field.}
```
```suggestion
\caption{\it Setup of the mid-ocean ridge model. Background colors show temperature. Black and white colors at the top of the model illustrate the orientation of the magnetic field frozen in the rock when the melt generated at the mid-ocean ridge reaches the surface, crystallizes to form new sea floor, and the rock cools down. White lines illustrate the flow field.}
```

naliboff commented on cookbooks/magnetic_stripes/magnetic_stripes.cc:
```diff
@@ -0,0 +1,119 @@
+    ASPECT_REGISTER_MATERIAL_MODEL(MagneticStripes,
+                                   "magnetic stripes",
+                                   "A simple material model that is like the "
+                                   "'melt simple' model, but has a constant reaction term.")
```
Can you add a bit more detail here about what the magnetic stripes represent and how they are calculated?

naliboff commented on cookbooks/magnetic_stripes/magnetic_stripes.cc:
```diff
@@ -0,0 +1,119 @@
+      for (unsigned int i=0; i < in.position.size(); ++i)
+        {
+          const double depth = this->get_geometry_model().depth(in.position[i]);
+          const double reaction_depth = 7000.0;
```
Add documentation regarding what this variable represents?

naliboff commented on cookbooks/magnetic_stripes/magnetic_stripes.cc:
```diff
@@ -0,0 +1,119 @@
+      for (unsigned int i=0; i < reversal_times.size(); ++i)
+        if (this->get_time() < reversal_times[i])
+          {
+            magnetic_orientation = -2 * (i % 2) + 1;
```
Add documentation about what this equation does? Does it simply produce alternating positive/negative values with each iteration?

naliboff commented on cookbooks/magnetic_stripes/magnetic_stripes.prm:
```diff
@@ -0,0 +1,250 @@
+# A model of a mid-ocean ridge that shows how magnetic lineations
```
Could this prm be simplified by referencing (i.e., "include") the previous mid-ocean ridge prm file and only specifying the modified or new sections?

jdannberg replied:
It probably could be, but since these are teaching cookbooks that someone should be able to follow without going through other cookbooks, I would prefer to not base it on other cookbooks.

naliboff replied:
Makes sense, happy to keep the PRM file in its current state.

naliboff commented on cookbooks/onset_of_convection/doc/onset_of_convection.tex:
```diff
@@ -0,0 +1,229 @@
+
+The input file for this model can be found at \url{cookbooks/onset_of_convection/onset_of_convection.prm}.
+
+This model is a  copy of the onset of convection benchmark described in Section~\ref{sec:benchmark-onset-of-convection}, 
```
```suggestion
This model is a copy of the onset of convection benchmark described in Section~\ref{sec:benchmark-onset-of-convection}, 
```

naliboff commented on cookbooks/onset_of_convection/doc/onset_of_convection.tex:
```diff
@@ -0,0 +1,229 @@
+What happens here is that the script performs a loop (a so-called for-loop):
+It executes all the commands within the loop for each of the values listed here in quotes. 
+In other words, this loop would run the model for all combinations of the viscosities $10^{24}$, $3.3 \cdot 10^{24}$, $10^{25}$, $3.3 \cdot 10^{25}$, $10^{26}$, $3.3 \cdot 10^{26}$, $10^{27}$ Pa s, and the temperature differences of 10, 30, 100, 300, 1000, 3000 and 10000 K between top and bottom of the model (which is what the DeltaT parameter stands for). 
+These are a lot of models, so these models are only run for 2 time steps each. 
```
```suggestion
Given the number of models, they are only run for 2 time steps each. 
```

naliboff commented on cookbooks/onset_of_convection/doc/onset_of_convection.tex:
```diff
@@ -0,0 +1,229 @@
+gnuplot plot.plt
+\end{verbatim}
+
+You can open \url{cookbooks/onset_of_convection/plot.plt} with leafpad to see what's in there, and you can also run each of these lines as a separate command interactively in in gnuplot. In order to do that, type \texttt{gnuplot} in the terminal, and then just copy or type any of the commands in the terminal window (but that's optional; hint: to close gnuplot, type \texttt{quit}). 
```
```suggestion
You can open \url{cookbooks/onset_of_convection/plot.plt} with leafpad to see what's in there, and you can also run each of these lines as a separate command interactively in gnuplot. In order to do that, type \texttt{gnuplot} in the terminal, and then just copy or type any of the commands in the terminal window (but that's optional; hint: to close gnuplot, type \texttt{quit}). 
```

naliboff commented on cookbooks/onset_of_convection/doc/onset_of_convection.tex:
```diff
@@ -0,0 +1,229 @@
+Then describe your model results. Include your plot of models with and without convection for different values of the viscosity and temperature difference and discuss it:
+What parameter range did you test?
+How do variations in each of the two parameters influence the occurrence of convection? 
+Describe how the line of the onset of convection (the interface between the part of your plot with and the one without convection)
```
```suggestion
Describe how the line of the onset of convection (i.e., the interface between the part of your plot with and the one without convection)
```

naliboff reviewed (commented):
@jdannberg - I've gone through all of the files, but I confess I lost momentum during the onset of convection section 😄. As noted in previous reviews, this is a fantastic contribution and thank you for all of the work that went into it!

I made quite a few comments, but all of them fall into the minor/stylistic category. In earlier sections I rewrote some of the text to use "you/your" less, but again lost steam doing this towards the end. In the manual, I thought less informal text would be better, but in retrospect I'm not sure that it really matters. If all of this is moved into Jupyter notebooks down the line, I think the current style would work well. Certainly nothing to overthink at this point or hold up merging the PR.

Thanks again and let me know if you have any questions or when this is ready for another review!

John

jdannberg commented on cookbooks/convection-box-particles/convection-box-particles.prm:
```diff
@@ -0,0 +1,158 @@
+
+
+# The next section deals with the initial conditions for the
+# temperature (there are no initial conditions for the
```
Since this file was copied from convection-box.prm, I am applying the corresponding changes to that file as well.

jdannberg commented on cookbooks/heat_flow/heat-flow-plume.prm:
```diff
@@ -0,0 +1,126 @@
+set Adiabatic surface temperature          = 1600
+
+############### Parameters describing the model
+# Let us here choose again a box domain of size 2x1
```
I don't know where this text first came from and what it referred to, but the identical text appears in the composition-reaction and platelike-boundary cookbooks. However, those have a nondimensional geometry, so rather than referring to them, I'm just removing that reference here (and fixing the numbers).

naliboff replied:
👍, thanks for fixing that!

jdannberg commented on cookbooks/heat_flow/heat-flow.prm:
```diff
@@ -0,0 +1,122 @@
+# A model of spreading at a mid-ocean ridge.
```
I copied the composition-reaction cookbook to use as a starting point, and that is also why it still has some of those old comments. But I think it does not actually make sense to read that first, since (1) it's less earth-like (nondimensional, in terms of boundary conditions, etc.) and it is more complicated in that it has the reaction part that is irrelevant here. So I am just going to try and remove all those references. I think since these are teaching cookbooks, they should be able to stand on their own anyway, and my students never looked at any of the other cookbooks before this one.

naliboff replied:
Agree that it is best to have the teaching cookbooks stand on their own (as much as possible). Thanks for going through and removing the old references/comments.

jdannberg commented:
@naliboff Thank you for your review! That must have been a lot of work. 
I think I addressed all of your comments. Let me know if you're happy with the new version, in that case I should probably rebase this and make sure everything still works with the current ASPECT version.

naliboff commented on cookbooks/magnetic_stripes/magnetic_stripes.cc:
```diff
@@ -113,7 +116,17 @@ namespace aspect
+                                   "degrees, the first compositional field used in the model "
+                                   "is set to either one or minus one. The sign depends on a "
+                                   "list of reversal times that is passed to the material "
+                                   "model as an input. Bevor the 1st, 3rd, 5th, ... entry, "
```
`Bevor` -> `Before`

jdannberg replied:
Thanks for catching this! That would have been the German way to write this, I was definitely too tired...

naliboff reviewed (commented):
@jdannberg - Thanks for addressing all of the comments, again for all of the work put into these cookbooks. I found one new minor typo, but otherwise I agree the next step should be to rebase and make sure everything runs properly. Perhaps worth adding simplified test cases for the new cookbooks to make sure they run with future changes? I am happy for that to come in a follow-up PR, so we can get this merged before the Hackathon.

jdannberg commented:
Okay, I fixed the typo and added the test cases. And it's good that I did, because for now they won't work, the magnetic stripes model needs #4586 so that the velocity is set correctly in the material model inputs. 

So we have to wait until that is merged, and then I can rebase again.

jdannberg commented:
I am not sure why the tests work (I got a floating point exception on my machine), but I guess that means that this is ready from my side.
